### ISSUE TITLE
Consider updating how fixed-sized buffers for non-primitive types are bound

### ISSUE THREAD

**Issue by tannergooding** [Ultz/Silk.NET#163]
Fixed sized buffers in C# only support a limited subset of types today: `bool`, `byte`, `char`, `short`, `int`, `long`, `sbyte`, `ushort`, `uint`, `ulong`, `float`, or `double`.

While there exists proposals for extending this to support other types as well, that support does not currently exist. This can make it difficult to work with certain structures such as [D3D12_RT_FORMAT_ARRAY](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_rt_format_array):
```cpp
struct D3D12_RT_FORMAT_ARRAY {
  DXGI_FORMAT RTFormats[8];
  UINT        NumRenderTargets;
};
```

Today, Silk.NET is generating an inline array of fields for these (similar to and based on https://discordapp.com/channels/521092042781229087/587346162802229298/696144071164428348):
```csharp
struct D3D12_RT_FORMAT_ARRAY
{
  public DXGI_FORMAT RTFormats_0;
  public DXGI_FORMAT RTFormats_1;
  public DXGI_FORMAT RTFormats_2;
  public DXGI_FORMAT RTFormats_3;
  public DXGI_FORMAT RTFormats_4;
  public DXGI_FORMAT RTFormats_5;
  public DXGI_FORMAT RTFormats_6;
  public DXGI_FORMAT RTFormats_7;
  public uint NumRenderTargets;
}
```

Instead, it may be possible to provide a "nicer" API by abusing some of the unsafe framework APIs:
```csharp
public partial struct D3D12_RT_FORMAT_ARRAY
{
    [NativeTypeName("DXGI_FORMAT [8]")]
    public _RTFormats_e__FixedBuffer RTFormats;

    [NativeTypeName("UINT")]
    public uint NumRenderTargets;

    public partial struct _RTFormats_e__FixedBuffer
    {
        internal DXGI_FORMAT e0;
        internal DXGI_FORMAT e1;
        internal DXGI_FORMAT e2;
        internal DXGI_FORMAT e3;
        internal DXGI_FORMAT e4;
        internal DXGI_FORMAT e5;
        internal DXGI_FORMAT e6;
        internal DXGI_FORMAT e7;

        public ref DXGI_FORMAT this[int index] => ref AsSpan()[index];

        public Span<DXGI_FORMAT> AsSpan() => MemoryMarshal.CreateSpan(ref e0, 8);
    }
}
```

This is abusing the `MemoryMarshal.CreateSpan` API to create a `Span` around the entire buffer and bypasses the normal C# language safety rules which would prevent this from occuring.

Because it is normally illegal C#, it is not recommended to expose the `ref` or `Span` as part of a "safe" API surface and it would be recommended to provide an analyzer that helps remind users that storing the `ref`/`Span` will not be tracked appropriately by the language and the `ref`/`Span` could outlive the actual backing struct if not managed appropriately.

It's also worth noting that the language could get proper support for this eventually, in which case the exposed signature may need to change slightly or in an incompatible way (depending on if they require any `modreq`/`modopt`, etc).

**Comment by Perksey** [comment]
CreateSpan not being available on .NET Standard 2.0 is a real bummer.

**Comment by tannergooding** [comment]
Slow span can't track interior pointers so it must be a pointer or array, which can already be handled via the exposed constructors.

**Comment by Perksey** [comment]
Resolved in #340. Example:
```cs
namespace Silk.NET.SDL
{
    [NativeName("Name", "SDL_MessageBoxColorScheme")]
    public unsafe partial struct MessageBoxColorScheme
    {
        
        [NativeName("Type", "SDL_MessageBoxColor [5]")]
        [NativeName("Type.Name", "SDL_MessageBoxColor [5]")]
        [NativeName("Name", "colors")]
        public ColorsBuffer Colors;

        public struct ColorsBuffer
        {
            public MessageBoxColor Element0;
            public MessageBoxColor Element1;
            public MessageBoxColor Element2;
            public MessageBoxColor Element3;
            public MessageBoxColor Element4;
            public ref MessageBoxColor this[int index]
            {
                get
                {
                    if (index > 4 || index < 0)
                    {
                        throw new ArgumentOutOfRangeException(nameof(index));
                    }

                    fixed (MessageBoxColor* ptr = &Element0)
                    {
                        return ref ptr[index];
                    }
                }
            }

#if NETSTANDARD2_1
            public Span<MessageBoxColor> AsSpan()
                => MemoryMarshal.CreateSpan(ref Element0, 5);
#endif
        }

    }
}
```
